nilearn/image/resampling.py

```diff
@@ -657,42 +545,235 @@ def resample_img(
-        # appears in padding images.
-        vmin = min(data.min(), 0)
-        vmax = max(data.max(), 0)
-        resampled_data.clip(vmin, vmax, out=resampled_data)
+    resampled_data = _clip_img(data, resampled_data, clip)
 
     return new_img_like(img, resampled_data, target_affine)
 
 
+def _get_resampled_datatype(interpolation, data, transform_affine):
+    """Cast unsupported data types to closest support dtype."""
+    if interpolation == "continuous" and data.dtype.kind == "i":
+        aux = data.dtype.name.replace("int", "float")
+        aux = aux.replace("ufloat", "float").replace("floatc", "float")
+        if aux in ["float8", "float16"]:
+            aux = "float32"
+        warnings.warn(f"Casting data from {data.dtype.name} to {aux}")
+        resampled_data_dtype = np.dtype(aux)
+    else:
+        resampled_data_dtype = data.dtype
+
+    # Since the release of 0.17, resampling nifti images have some issues
+    # when affine is passed as 1D array
+    # and if data is of non-native endianness.
+    # See issue https://github.com/nilearn/nilearn/issues/1445.
+    # If affine is passed as 1D, scipy uses _nd_image.zoom_shift rather
+    # than _geometric_transform (2D) where _geometric_transform is able
+    # to swap byte order in scipy later than 0.15 for nonnative endianness.
+
+    # We convert to 'native' order to not have any issues either with
+    # 'little' or 'big' endian data dtypes (non-native endians).
+    A, _ = to_matrix_vector(transform_affine)
+    if len(A.shape) == 1 and not resampled_data_dtype.isnative:
+        resampled_data_dtype = resampled_data_dtype.newbyteorder("N")
+
+    return resampled_data_dtype
+
+
+def _get_bounding_box(target_affine, target_shape, affine, data):
+    """Get a bounding box for the transformed data.
+
+    If necessary:
+        - embed target_affine in 4x4 shape
+        - calculate offset from bounding box
+        - compute target_shape
+
+    Parameters
+    ----------
+    target_affine : numpy.ndarray
+        If specified, the image is resampled corresponding to this new affine.
+        target_affine can be a 3x3 or a 4x4 matrix.
+
+    target_shape : tuple or list
+        If specified, the image will be resized to match this new shape.
+        len(target_shape) must be equal to 3.
+        If target_shape is specified, a target_affine of shape (4, 4)
+        must also be given.
+
+    affine : numpy.ndarray
+        Affine of the input image.
+
+    data : numpy.ndarray
+        data of the input image.
+    """
+    if target_affine.shape == (3, 3):
+        # Embed target_affine in 4x4 shape
+        missing_offset = True
+        target_affine_tmp = np.eye(4)
+        target_affine_tmp[:3, :3] = target_affine
+        target_affine = target_affine_tmp
+    else:
+        missing_offset = False
+        target_affine = target_affine.copy()
+
+    transform_affine = np.linalg.inv(target_affine).dot(affine)
```

Comment:
this seems to be the inverse of the `transform_affine` defined in `resample_img`? only the shape of the `transform_affine` in `resample_img` matters, but it's a bit confusing they have the same name